### PR TITLE
[#3038] Add buffer to process incoming messages asynchronously

### LINKED ISSUE #2852: [Perf] User story 1: Time lags navigating through different app views after log in when user has a lot of sent/received messages
***Type***: Bug

***Summary***: When user logs in into account, application becomes unresponsive for some time. It depends from how many messages user sent or received - more messages cause more time for app to hung up.
It's also visible that application retrieves all the messages user sent/received in chats it had before log out.
For example, in my case with Samsung J7  I had user with about 150 received messages. With such amount, when I logged out and logged back in - it took ~3 minutes since I could with no lags navigate through different views (chats, Profile, Wallet etc). Below TF session session attached.
On video it's especially visible when Profile screen hides during 01:30-02:00.

On iPhone X it's also visible (50 sent messages is enough to have before log in to account) 

TF logs: https://app.testfairy.com/projects/4803622-status/builds/7397034/sessions/4/?accessToken=DCNt-8FbqXifdFksvbCKKRoFwlw

### Expected behavior
No time lags while navigating through different app screens after user logged in

### Actual behavior
Time lags while navigating through different app screens after user logged in

### Reproduction
**Precondition**: User exist and has more than 100 messages sent/received (actually it's even enough 20 messages - depends from device)

- Open Status
- Log in to account
- Navigate through different screen (Profile, 1-1 chat, Wallet, etc)

### Additional Information
* Status version: Develop 0.9.10-493-g77b966c8+
* Operating System: Android 7.0 (Samsung J7 real device), iOS 11.1.2 (iPhone X)

### PR DESCRIPTION
**Pull request by dmitryn**
Experiment around #3038 
Addresses #2852 

### Summary:

[comment]: # (Summarise the problem and how the pull request solves it)

Process events related to incoming messages asynchronous using buffer. This allows other events (UI-related) to be dispatched and processed so app is not freezing when getting a lot of chat messages.

### Testing notes (optional):
* App with a lot of messages in chats should not freeze on start.
* Chats with a lot of messages should not freeze UI.

status: ready

### CODE REVIEW

**Comment by janherich** [comment]
@dmitryn looks interesting, what are the performance results of such changes ? Also I don't think we actually want to use sliding buffer, even if the buffer size is quite high (1024), so it's super unlikely that we would fill it, in the case we do, I'm sure we would rather accept some slowdown rather then drop events like `:incoming-message`, which would hide the updates (new message, status, etc.) from user.
So instead of sliding buffer, I think that regular fixed buffer (which will block and wait when it's full) is better for worst-case scenarios. 
Also another thing, did you try to just move all potentially longer running actions (like realm updates) into async actions and leaving events as they are ?

**Comment by dmitryn** [comment]
@janherich

>Also I don't think we actually want to use sliding buffer, even if the buffer size is quite high (1024), so it's super unlikely that we would fill it

If you fill fixed `chan` with > 1024 messages you still get `No more than 1024 pending puts are allowed on a single channel. Consider using a windowed buffer.`
https://github.com/clojure/core.async/blob/0498ba69fda5d930e3e7568cc90ca933c19c42ca/src/main/clojure/cljs/core/async/impl/channels.cljs#L86  and app will crash. But i agree, it's highly unlikely case.

>did you try to just move all potentially longer running actions (like realm updates) into async actions and leaving events as they are ?

No, as it's not a part of this PR. Current PR solves issue with re-frame event processing where queue gets overwhelmed and new events doesn't get processed until previous processing isn't done https://github.com/Day8/re-frame/blob/master/src/re_frame/router.cljc#L198

**Comment by dmitryn** [comment]
@janherich 
>what are the performance results of such changes ?

I don't have specific numbers. But in general, app doesn't freeze anymore when you start it and have a lot of incoming messages. UI works so you can actually browse the app.

**Review by janherich — APPROVED** [review]

File: src/status_im/utils/events_buffer.cljs [hunk 1]
```diff
@@ -3,10 +3,21 @@
             [re-frame.core :as re-frame])
   (:require-macros [cljs.core.async.macros :refer [go-loop]]))
 
-(defonce ^:private buffer (async/chan (async/sliding-buffer 1024)))
-
+;; NOTE:(dmitryn) Using buffer here because using channel
+;; raises when there are more than 1024 pending puts.
+;; Ideally we should not exceed current buffer size.
+;; https://github.com/clojure/core.async/blob/0498ba69fda5d930e3e7568cc90ca933c19c42ca/src/main/clojure/cljs/core/async/impl/channels.cljs#L86
+(defonce ^:private buffer (async/chan (async/buffer 10000)))
```
Comment (janherich) [inline src/status_im/utils/events_buffer.cljs]: Just a small hint, the above can be more succinctly written as `(async/chan 1000)` - whenever you specify buffer size as an integer, fixed buffer is created by default.

**Review by oskarth — APPROVED** [review]
Good idea

**Comment by Serhy** [comment]
This has been tested with the user having 1000+ read messages in 50 different threads. 

1) Using application of release version 0.9.12+ -> delay when navigating through different screens (Chats -> Wallet -> Profile -> Chats -> Add new contact -> Back to Chats screen -> ...), after log in were up to 5 seconds during first minute.

2) Same application with 0.9.12 release version upgraded to the 8ad238bf+ app version + (this build with enabled 'Offline Inbox' flag) and tested in the same way (Log in to account -> navigate through different screens). 
Delay when navigating through different screens (Chats -> Wallet -> Profile -> Chats -> Add new contact -> Back to Chats screen - ...) was seen up to 10 seconds during first minute.
(See TF session - https://app.testfairy.com/projects/4803622-status/builds/7513158/sessions/11/?accessToken=ZrakE2SMFqjVogjZt5nuM0eYy0k from 00:00 to 02:00)

3) Same application with 0.9.12 release version upgraded to the 0.9.10-585-g8ad238bf+ app version (build with disabled 'Offline Inbox' flag) and tested in the same way (Log in to account -> navigate through different screens) 
Delay when navigating through different screens (Chats -> Wallet -> Profile -> Chats -> Add new contact -> Back to Chats screen - ...) noticed up to 4-5 seconds
TF session: https://app.testfairy.com/projects/4803622-status/builds/7520855/sessions/1/?accessToken=bOuQO3oBrWRXfP/jIc0ahtWPsCo

4) Same 0.9.12 App upgraded to latest 0.9.13 Release (64th release from 16th of January): 
After I logged in to account - since 1 minute app froze up and crashed (no error - just blank screen) https://app.testfairy.com/projects/4803622-status/builds/7513197/sessions/22/?accessToken=p59nkHBwlkFEKorIUk03mHK7UGU

**Comment by Serhy** [comment]
I also don't see any regressions. 
Lets merge changes made in this PR. In separate issue we'll address further improvements for the #2852.
Nice job, @dmitryn ! Thanks a lot!

**Pull request merged**